compiler/src/dotty/tools/dotc/transform/init/Objects.scala

```diff
@@ -0,0 +1,1419 @@
+            val addr = ref.varAddr(target)
+            if addr.owner == State.currentObject then
+              Heap.read(addr)
+            else
+              errorReadOtherStaticObject(State.currentObject, addr.owner)
+              Bottom
+          else if ref.isObjectRef && ref.klass.hasSource then
+            report.warning("Access uninitialized field " + field.show + ". Call trace: " + Trace.show, Trace.position)
+            Bottom
+          else
+            // initialization error, reported by the initialization checker
+            Bottom
+        else if ref.hasVal(target) then
+          ref.valValue(target)
+        else if ref.isObjectRef && ref.klass.hasSource then
+          report.warning("Access uninitialized field " + field.show + ". Call trace: " + Trace.show, Trace.position)
+          Bottom
+        else
+          // initialization error, reported by the initialization checker
+          Bottom
+
+      else
+        if ref.klass.isSubClass(receiver.widenSingleton.classSymbol) then
+          report.error("[Internal error] Unexpected resolution failure: ref.klass = " + ref.klass.show + ", field = " + field.show + Trace.show, Trace.position)
+          Bottom
+        else
+          // This is possible due to incorrect type cast.
+          // See tests/init/pos/Type.scala
+          Bottom
+
+    case fun: Fun =>
+      report.error("[Internal error] unexpected tree in selecting a function, fun = " + fun.code.show + Trace.show, fun.code)
+      Bottom
+
+    case Bottom =>
+      if field.isStaticObject then ObjectRef(field.moduleClass.asClass)
+      else Bottom
+
+    case RefSet(refs) =>
+      refs.map(ref => select(ref, field, receiver)).join
+  }
+
+  /** Handle assignment `lhs.f = rhs`.
+   *
+   * @param lhs         The value of the object to be mutated.
+   * @param field       The symbol of the target field.
+   * @param lhs         The value to be assigned.
+   * @param rhsTyp      The type of the right-hand side.
+   */
+  def assign(lhs: Value, field: Symbol, rhs: Value, rhsTyp: Type): Contextual[Value] = log("Assign" + field.show + " of " + lhs.show + ", rhs = " + rhs.show, printer, (_: Value).show) {
+    lhs match
+    case fun: Fun =>
+      report.error("[Internal error] unexpected tree in assignment, fun = " + fun.code.show + Trace.show, Trace.position)
+
+    case Cold =>
+      report.warning("Assigning to cold aliases is forbidden", Trace.position)
+
+    case Bottom =>
+
+    case RefSet(refs) =>
+      refs.foreach(ref => assign(ref, field, rhs, rhsTyp))
+
+    case ref: Ref =>
+      if ref.hasVar(field) then
+        val addr = ref.varAddr(field)
+        if addr.owner != State.currentObject then
+          errorMutateOtherStaticObject(State.currentObject, addr.owner)
+        else
+          Heap.write(addr, rhs)
+      else
+        report.warning("Mutating a field before its initialization: " + field.show, Trace.position)
+    end match
+
+    Bottom
+  }
+
```

Comment:
Does assign actually need an abstract value? In my intuition assign statements do not yield values, and it seems that all it returns here is `Bottom`. Maybe we can change it to `Contextual[Unit]`